### DISCUSSION TITLE
Nesting definitions.

### DISCUSSION THREAD

**Discussion by SirWumpus** [ForthHub/discussion#159] [Q&A] [answered]
I've had a GitHub issued opened concerning [Post4](https://github.com/SirWumpus/post4/issues/4).  At the heart of it is the issue of nesting definitions.  This code from [Rosetta](https://rosettacode.org/wiki/Function_composition#Forth):

```
: compose ( xt1 xt2 -- xt3 )
  >r >r :noname
     r> compile,
     r> compile,
     postpone ;
;
```

I contend this is invalid according to the Forth 200x draft 19.1:

> 3.4.5 Compilation
> 
> A program shall not attempt to nest compilation of definitions.
>
> During the compilation of the current definition, a program shall not execute any defining word, :NONAME, or any definition that allocates dictionary data space. The compilation of the current definition may be suspended using [ (left-bracket) and resumed using ] (right-bracket). While the compilation of the current definition is suspended, a program shall not execute any defining word, :NONAME, or any definition that allocates dictionary data space.

* Am I misinterpreting the 200x draft?
* Was this valid against the older 1994 standard?

**Comment by ruv** [comment, marked as answer]
This definition for `compose` does not contain nested definitions, and there is no attempt to nest compilation of definitions. So, this definition for `compose` is standard-compliant in Forth-1994, as well as in Forth-2012.

A problem can occur when executing `compose`.  Namely, if it is executed while another definition is being compiled (i.e., _the current definition_ is present), it violates 3.4.5. 

A correct usage example:
```forth
' foo ' bar compose constant func-foobar
```
In this case, the current definition is absent when `compose` is executed.

An incorrect usage example:
```
: baz  [ ' foo  ' bar compose compile, ] ;
``` 
In this case, the current definition is present when `compose` is executed — namely, compilation for `baz` has been started and is not yet finished.

**Reply by SirWumpus** [reply]
@ruv Thank you for quick response and clarity.
